torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import random
+from typing import Dict, Optional, TypeVar, Union
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+T = TypeVar("T")
+
+
+@functional_datapipe("random_split")
+class RandomSplitterIterDataPipe(IterDataPipe):
+    r"""
+    Randomly split samples from a source DataPipe into groups(functional name: ``random_split``).
+    Since there is no buffer, only ONE group of samples (i.e. one child DataPipe) can be iterated through
+    at any time. Attempts to iterate through multiple of them simultaneously will fail.
+
+    Note that by default, multiple iterations of this DataPipe will yield the same split for consistency across epochs.
+    You can invoke ``set_seed`` on the output(s) to update the seed whenever needed (such as per epoch).
+
+    Args:
+        source_datapipe: Iterable DataPipe being split
+        total_length: Length of the ``source_datapipe``, optional but providing an integer is highly encouraged,
+            because not all ``IterDataPipe`` has ``len``, espeically ones that can be easily known in advance.
+        weights: Dict of weights; the length of this list determines how many output DataPipes there will be.
+        seed: random _seed used to determine the randomness of the split
+        target: Optional key (that must exist in ``weights``) to indicate the specific group to return.
+            If set to the default ``None``, returns ``List[IterDataPipe]``.
+            If target is specified, returns ``IterDataPipe``.
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper(range(10))
+        >>> train, valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0)
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+        >>> # You can also specify a target key if you only need a specific group of samples
+        >>> train = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='train')
+        >>> list(train)
+        [2, 3, 5, 7, 8]
+        >>> # Be careful to use the same seed as before when specifying `target` to get the correct split.
+        >>> valid = dp.random_split(total_length=10, weights={"train": 0.5, "valid": 0.5}, seed=0, target='valid')
+        >>> list(valid)
+        [0, 1, 4, 6, 9]
+    """
+
+    def __new__(
+        cls,
+        source_datapipe: IterDataPipe,
+        weights: Dict[T, Union[int, float]],
+        seed,
```

Comment:
I have been thinking about that and I'm on the fence about what makes sense given the use case of `random_split` is different from `shuffle` in that most users probably want to pick one static `seed` and that forget about it for `random_split`. Whereas it is more important to control randomness for `shuffle`.

If we choose to remove `seed` from the argument, and the `seed` is `None`, would you suggest initializing with `self._seed = int(torch.empty((), dtype=torch.int64).random_().item())`?

I am inclined not to do that, and just initialize `random.Random()` with `None` so that the default seed isn't different between runs.

Reply:
Related to the [comment](https://github.com/pytorch/data/pull/724/files#r948363330). It depends on how you want to do this split.
If you think it's more like a static split, I am fine to keep seed as an argument. But, I think `set_seed` API should be removed from those DataPipe to prevent `DataLoader2` automatically reset seed every time. You can imagine this would cause splits are not mutually exclusive and collectively exhaustively with `train_dataloader`, `test_dataloader` and `valid_dataloader`.

Reply:
I am thinking the split should be "mostly static, with optional `set_seed` for special usages".

I think it would be nice to have an option to replace the `seed` but I don't feel strongly enough about it. Maybe remove the method `set_seed` is the best option for now.

cc: @VitalyFedyunin Any thoughts?

Reply:
I supposed there might be use cases where users just want to randomly sample x% of data with this DataPipe? In that use case they would want things to be different per epoch.

How about I add an argument `fixed_seed: bool = True`. If `True`, `set_seed` doesn't change the seed. If `False`, it does change.

Reply:
For the case of `fixed_seed = False`, I personally think users would want to have fixed splits within each epoch but different seeds for every epoch.

It becomes tricky for one existing example from pytorch lighting that they need multiple DataLoaders per epoch. But, each `DataLoader` would reset the seed per iteration. Then, we have to ask users to manually reset global RNG after iteration of each dataloader. Code as shown below:
```py
train_dp, val_dp = dp.random_split(total_length=10, weights={"train": 0.7, "valid": 0.3}, seed=0)
train_dataloader = DataLoader(train_dp)
val_dataloader = DataLoader(val_dp)

for epoch in range(3):
    torch.manual_seed(epoch)
    for d in train_dataloader:  # New seed for random_split
        model(d)

    torch.manual_seed(epoch)
    for d in val_dataloader:  # Same seed for random_split
        model(d)
```
`torch.manual_seed(epoch)` has to be placed before the iteration of each DataLoader to make sure the splits static within one epoch, because `apply_seed` will be invoked by `DataLoader2` at the beginning of each iteration.

Reply:
I was not expecting users to do `train` `val` split with `fixed_seed = False`, because I don't think they would want to use a different training set for every epoch, right?

I was thinking that if they use `fixed_seed = False`, they likely just want to randomly sample say 50% of the training data, which can be different per epoch and the rest will be dropped.

If my thinking is incorrect, then we should drop the argument `fixed_seed` entirely and go with the seed being static?

Reply:
> I was thinking that if they use `fixed_seed = False`, they likely just want to randomly sample say 50% of the training data, which can be different per epoch and the rest will be dropped. Perhaps we should just separate these two functionalities entirely.

I feel this use case is more like a sampling rather than splitting.

Reply:
I removed `fixed_seed` as well as renamed `set_seed` then. The intention here is that the `seed` should be static across epochs because I don't believe users want to change the seed by epoch (let me know if this is not true). We will allow `override_seed` but generally we don't expect users to use it.

Any comments?

Reply:
Sounds good. In terms of the name, could we do something including `epoch` to give users a hint to only call it per epoch?

Reply:
I am hesitate to do that in the name since technically they can use the method separately (not necessarily tied to epochs), but I will update the docstring to indicate that may be a common use case.